pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(100)]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			tranches: Vec<(u8, u8)>,
+			currency: T::CurrencyId,
+			max_reserve: T::Balance,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+
+			// TODO: Ensure owner is authorized to create a pool
+
+			// A single pool ID can only be used by one owner.
+			ensure!(!Pool::<T>::contains_key(pool_id), Error::<T>::PoolInUse);
+
+			// At least one tranch must exist, and the last
+			// tranche must have an interest rate of 0,
+			// indicating that it recieves all remaining
+			// equity
+			ensure!(
+				tranches.last() == Some(&(0, 0)),
+				Error::<T>::NoJuniorTranche
+			);
+
+			let now = T::Time::now().as_secs();
+			let tranches = tranches
+				.into_iter()
+				.map(|(interest, sub_percent)| {
+					const SECS_PER_YEAR: u64 = 365 * 24 * 60 * 60;
+					let interest_per_sec =
+						Perquintill::from_percent(interest.into()) / SECS_PER_YEAR;
+					Tranche {
+						interest_per_sec,
+						min_subordination_ratio: Perquintill::from_percent(sub_percent.into()),
+						epoch_supply: Zero::zero(),
+						epoch_redeem: Zero::zero(),
+
+						debt: Zero::zero(),
+						reserve: Zero::zero(),
+						ratio: Perquintill::zero(),
+						last_updated_interest: now,
+					}
+				})
+				.collect();
+			Pool::<T>::insert(
+				pool_id,
+				PoolDetails {
+					owner: owner.clone(),
+					currency,
+					tranches,
+					current_epoch: One::one(),
+					last_epoch_closed: now,
+					last_epoch_executed: Zero::zero(),
+					closing_epoch: None,
+					max_reserve,
+					available_reserve: Zero::zero(),
+					total_reserve: Zero::zero(),
+				},
+			);
+			Self::deposit_event(Event::PoolCreated(pool_id, owner));
+			Ok(())
+		}
+
+		#[pallet::weight(100)]
+		pub fn order_supply(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			tranche_id: T::TrancheId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// TODO: Ensure this account is authorized for this tranche
+			let (currency, epoch) = {
+				let pool = Pool::<T>::try_get(pool_id).map_err(|_| Error::<T>::NoSuchPool)?;
+				ensure!(pool.closing_epoch.is_none(), Error::<T>::PoolClosing);
+				(pool.currency, pool.current_epoch)
+			};
+			let tranche = TrancheLocator {
+				pool_id,
+				tranche_id,
+			};
+			let pool_account = PoolLocator { pool_id }.into_account();
+
+			Order::<T>::try_mutate(&tranche, &who, |order| -> DispatchResult {
+				if amount > order.supply {
+					let transfer_amount = amount - order.supply;
+					Pool::<T>::try_mutate(pool_id, |pool| {
+						let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+						let epoch_supply = &mut pool.tranches[tranche_id.into()].epoch_supply;
+						*epoch_supply = epoch_supply
+							.checked_add(&transfer_amount)
+							.ok_or(Error::<T>::Overflow)?;
+						T::Tokens::transfer(currency, &who, &pool_account, transfer_amount)
```

Comment:
Can this fail? Mustn't we have an early return here too? Otherwise, `T::Tokens::transfer` might fail, but we still insert `Order`?